pkg/reconciler/ingress/ingress.go

```diff
@@ -0,0 +1,279 @@
+		// a cluster.
+		//
+		// This update can come from the creation or because the syncer has update the status.
+
+		rootIngressName := ingress.Labels[ownedByLabel]
+		// A leaf Ingress was updated; get others and aggregate status.
+		sel, err := labels.Parse(fmt.Sprintf("%s=%s", ownedByLabel, rootIngressName))
+		if err != nil {
+			return err
+		}
+		others, err := c.lister.List(sel)
+		if err != nil {
+			return err
+		}
+
+		// Get the rootIngress based on the labels.
+		var rootIngress *networkingv1.Ingress
+
+		rootIf, exists, err := c.indexer.Get(&networkingv1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:   ingress.Namespace,
+				Name:        rootIngressName,
+				ClusterName: ingress.GetClusterName(),
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		// TODO(jmprusi): A leaf without rootIngress?
+		if !exists {
+			return fmt.Errorf("Root Ingress not found: %s", rootIngressName)
+		}
+
+		rootIngress = rootIf.(*networkingv1.Ingress).DeepCopy()
+
+		// Clean the current status, and then recreate if from the other leafs.
+		rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{}
+		for _, o := range others {
+			rootIngress.Status.LoadBalancer.Ingress = append(rootIngress.Status.LoadBalancer.Ingress, o.Status.LoadBalancer.Ingress...)
+		}
+
+		// If the envoy controlplane is enabled, we update the cache and generate and send to envoy a new snapshot.
+		if c.envoycontrolplane != nil {
+			err := c.envoycontrolplane.UpdateIngress(*rootIngress)
+			if err != nil {
+				return err
+			}
+
+			statusHost := generateStatusHost(c.domain, rootIngress)
+			// Now overwrite the Status of the rootIngress with our desired LB
+			rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{{
+				Hostname: statusHost,
+			}}
+		}
+
+		// Update the rootIngress status with our desired LB.
+		if _, err := c.client.NetworkingV1().Ingresses(rootIngress.Namespace).UpdateStatus(ctx, rootIngress, metav1.UpdateOptions{}); err != nil {
+			if errors.IsConflict(err) {
+				key, err := cache.MetaNamespaceKeyFunc(ingress)
+				if err != nil {
+					return err
+				}
+				c.queue.AddRateLimited(key)
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+// desiredLeaves returns a list of leaves (ingresses) to be created based on an ingress.
+func (c *Controller) desiredLeaves(ctx context.Context, root *networkingv1.Ingress) ([]*networkingv1.Ingress, error) {
+	// This will parse the ingresses and extract all the destination services,
+	// then create a new ingress leaf for each of them.
+	services, err := c.getServices(ctx, root)
+	if err != nil {
+		return nil, err
+	}
+
+	var clusterDests []string
+	for _, service := range services {
+		if service.Labels[clusterLabel] != "" {
+			clusterDests = append(clusterDests, service.Labels[clusterLabel])
+		} else {
+			klog.Infof("Skipping service %q because it is not assigned to any cluster", service.Name)
```

Comment:
Do you need a `continue` here? Or is it appropriate to add this service to the tracker even though it's not labeled?